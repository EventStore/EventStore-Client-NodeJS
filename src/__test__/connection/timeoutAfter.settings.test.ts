import { Channel } from "@grpc/grpc-js";

import { EventStoreDBClient } from "../..";

/*
Mocking this file breaks grpc, but allows us to check the settings passed to grpc
These tests need to be kept seperate to any tests that need to actually make calls 
*/
jest.mock("@grpc/grpc-js/build/src/channel.js");
const ChannelMock = Channel as jest.Mock<Channel>;

describe("timeoutAfter settings", () => {
  beforeEach(() => {
    ChannelMock.mockClear();
  });

  describe.each([
    [
      "should default to 5_000", // test name
      "esdb://host", // connection string
      {}, // constructor options
      undefined, // call options
      5_000, // expected
    ],
    [
      "should be settable: 1",
      "esdb://host?timeoutAfter=1",
      { timeoutAfter: 1 },
      undefined,
      1,
    ],
    [
      "should be settable: 100000",
      "esdb://host?timeoutAfter=100000",
      { timeoutAfter: 10_0000 },
      undefined,
      10_0000,
    ],
    [
      "passing in call options should override default",
      "esdb://host",
      {},
      { timeoutAfter: 10_0000 },
      10_0000,
    ],
    [
      "passing in call options should override settings",
      "esdb://host?timeoutAfter=100000",
      { timeoutAfter: 10_0000 },
      { timeoutAfter: 10 },
      10,
    ],
  ])("%s", (_, connectionString, constructorOptions, callOptions, expected) => {
    test.each([
      [
        "connectionString",
        () => EventStoreDBClient.connectionString(connectionString),
      ],
      [
        "constructor",
        () =>
          new EventStoreDBClient({
            endpoint: "host:1234",
            ...constructorOptions,
          }),
      ],
    ])("%s", async (_, createClient) => {
      const warnSpy = jest.spyOn(console, "warn").mockImplementation();
      const client = createClient();
      const before = Date.now();

      try {
        await client.restartSubsystem(callOptions);
      } catch (_) {
        // We're not actually connecting to anything, just triggering channel creation
      }

      const after = Date.now();

      const ChannelInstance = ChannelMock.mock.instances[0];
      const createCall = (ChannelInstance.createCall as unknown) as jest.Mock<
        Channel["createCall"]
      >;

      const deadline = createCall.mock.calls[0][1];

      expect(deadline).toBeGreaterThanOrEqual(before + expected);
      expect(deadline).toBeLessThanOrEqual(after + expected);

      expect(warnSpy).not.toBeCalled();
      warnSpy.mockRestore();
    });
  });

  describe.each([
    ["should throw on zero", "esdb://host?timeoutAfter=0", { timeoutAfter: 0 }],
    [
      "should throw on negative",
      "esdb://host?timeoutAfter=-1",
      { timeoutAfter: -1 },
    ],
    [
      "should throw on negative",
      "esdb://host?timeoutAfter=-1000000000000000",
      { timeoutAfter: -1000000000000000 },
    ],
  ])("%s", (_, connectionString, constructorOptions) => {
    test.each([
      [
        "connectionString",
        () => EventStoreDBClient.connectionString(connectionString),
      ],
      [
        "constructor",
        () =>
          new EventStoreDBClient({
            endpoint: "host:1234",
            ...constructorOptions,
          }),
      ],
    ])("%s", async (_, createClient) => {
      expect(() => createClient()).toThrowErrorMatchingSnapshot();
    });
  });
});
