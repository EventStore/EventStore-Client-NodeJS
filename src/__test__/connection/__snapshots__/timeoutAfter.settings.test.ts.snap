// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeoutAfter settings should throw on negative connectionString 1`] = `"Invalid timeoutAfter \\"-1\\". Please provide a positive integer."`;

exports[`timeoutAfter settings should throw on negative connectionString 2`] = `"Invalid timeoutAfter \\"-1000000000000000\\". Please provide a positive integer."`;

exports[`timeoutAfter settings should throw on negative constructor 1`] = `"Invalid timeoutAfter \\"-1\\". Please provide a positive integer."`;

exports[`timeoutAfter settings should throw on negative constructor 2`] = `"Invalid timeoutAfter \\"-1000000000000000\\". Please provide a positive integer."`;

exports[`timeoutAfter settings should throw on zero connectionString 1`] = `"Invalid timeoutAfter \\"0\\". Please provide a positive integer."`;

exports[`timeoutAfter settings should throw on zero constructor 1`] = `"Invalid timeoutAfter \\"0\\". Please provide a positive integer."`;
